### ISSUE TITLE
Create Threaded and Scaleable General Architecture draft

### ISSUE THREAD

**Issue by a-teammate** [inexor-game/code#321]
Currently Inexor is mainly single threaded and (in most places) blocking and synchronous.  
Only exceptions are the sound backend which (thanks to SDL) plays in another thread and accepts commands in a nonblocking fashion.
Also the rendering is (thanks to the OpenGL drivers) at many places asynchronous _(Note: but still blocking by times: while some calls are nonblocking, follow up commands using the result of the previous could still block and wait for the result before)_.  

Since we will move a lot of code in the next 2 releases (into modules, namespaces and classes) we could also already work towards a better future in CPU load way.  

So what we need right now is a working concept of a nicely scale-able Architecture, where jobs get executed by a threadpool, things are balanced on the cores nicely and **most importantly:** the **communication between** the different next to each other running **systems** is **save** and **fast** (no useless communication at best).  

At a starting point we should start small with a new application (just sth to test and does not really be related) which provides good solutions for these problems:  
* **organizing modules and their jobs**
   * and maybe modules are just a bunch of smaller jobs which could work in parallel too?
       * like the particle system which has a lot of small particles and modifiers which influence these particles
   * some jobs can only run on one thread
     * (e.g. all rendering calls have to happen in one consistent thread)
   * do things get chained? or is a manager saying: physics system do a tick(), rendering do 2 ticks().. ? 
* **communication between those**
   * some modules have an input(s) from one module(s) and an output to another 
       * physics take input e.g. from the entity system (e.g. players position) and output to it (the next player position)
       * the entity system should share the new position over the network for the other players
           * and receive as well ofc
    * waiting times for input should be small
      * best case ofc not existent 
    * no useless work should be done though (except if thats faster)
       * some jobs have fixed rates (e.g. physics system updates only every 33ms)
    * communication in fixed (or better clamped) rates could also be good sometimes
       * e.g. a gravity field (beeing a particle) flying through the air doesn't need to share its new position everytime it calculates a new one

These we're just random aims I wrote down here, some may not really be the aim for the test-app but it would be good if those could be done with the general system.

Libs which could be helpful:  
* https://github.com/chriskohlhoff/executors
   * alternatives see #160 and maybe new ones appeared? what happened to that proposal?
* boost signals2
   * the Observer pattern in general:
      * maybe most syncing can be done automatically when you change a variable
          * -> everyone else needing this variable gets notified
              * watch out for deadlocks: A -> B (varxy modified) .. B-> A (varxy modified) ... A -> B.....
      * checkout inexor/utils/Observe.hpp
* maybe queues sometimes
   * best nonblocking, but you also dont want a job if theres no work..
* .. (maybe you find some more awesome concepts or patterns)

I'd like to think of this as a big construction side (or sth like that):

The People creating the windows need wood from the wood-creating people.
More then one window can be created (one person per window? or maybe more people?)
Changes in the design of the windows should be told to the wood people (by the design guys).
Also it should be told to the family that will live in the house. Or maybe that family is requesting that change. -> Everyones working in parallel, having their jobs and information (and requests)... but also people are limited ofc (and only some can do a special job, but there are a lot of jobs everyone can do). Maybe there are meetings? Do there need to be meetings?  
Well that is just what helps me imagining that before my inner eye, dunno if it does help you.

Maybe there are already living proof-of-concepts out there?( Creating our own won't hurt in any case though.)

@impowski maybe you could look into this? I had the feeling you are interested in this topic and it would be god damn helpful (even if it answers the problems just partly) for the next steps which involve refractoring the code in the direction of that draft.

**Comment by impowski** [comment]
@a-teammate Yeah, I should look at this, sometimes I just get frustrated by complexity of the code. I don't get this part
> At a starting point we should start small with a new application
What do you mean by that? So basically as I understood you want a better CPU management so that nothing will stuck on same thread with something else? I was working with signals in Qt but I mean I need a place to start from, because as I said before some of parts in code are really confusing sometimes I miss the idea in architecture.

**Comment by a-teammate** [comment]
_**Update:**_ 
### Compontents
We will work towards a modular system using a component-based framework.
(thats basically an even more drastic approach than solely object orientated programming, see [wikipedia](https://en.wikipedia.org/wiki/Component-based_software_engineering)).

It means dividing the code into a lot of subsystems which act mostly on their own (but can communicate with each other when they subscribe to another subsystem).
We will probably use [cppmicroservices](http://cppmicroservices.org/) as our lib in that regard.

That already means a clarified look (interface) of subsystems and using cppmicroservices the communication between those components is based on "hey i subscribe that, tell me when it changes". 

### Communication

Dealing with threading (and scale-ability) the hardest you have to deal with is communication between threads.
There are mainly two approaches: shared data (which can be accessed from both threads, but are saved from becoming corrupted by locks) and on the other side message based approaches (where data is never shared but every thread has its data and gets notified about updates on this data using some kind of messaging). (Okay there are some more like transactional memory, which is not really production ready yet though).

Okay so we choose a message driven approach over the shared data since shared data is really error prone.
One popular approach for such communication through messages is the actor model (which adds that the messages should be send asynchronously).
That guarantee of asynchronously (nonblocking) communication is however not given by cppmicroservices (I assume at least), so its not really an actor framework. (And it would probably still require some effort in those subscriptions callbacks to make them thread-safe and safely interruptible).

Consequently the next step will then be to add another actor framework which handles exactly that!
Either we build our own (maybe using boost::signals2 combined with some executor (executor=schedules jobs to specific threads)), or maybe even better using the [c++actor-framework](http://actor-framework.readthedocs.io/en/latest/index.html).

So i think we can close this, since we don't really need a standalone demo app anymore (We could simply separate one subsystem after the other with component based programming).  
But yeah @impowski thank you for reading into this super advanced topic and thank you @aschaeffer for finding the cppmicroservices lib!
I think we can now make the new design clearer step by step.

**Comment by impowski** [comment]
@a-teammate I like CAF but I think it might be a far a lot of stuff for this project and we don't know anything about it's stability. Microservices is something I think we can not use but use a better approach some sort of Design Pattern maybe, because we are might get deeper into OOP and lose our minds somewhere.

**Comment by a-teammate** [comment]
Yes true, a good design is independent of its underlying library:

If we have independent components (with a defined interface), a defined communication interface between them and defined interface for a registry for all components, the deployment can be anything (we can deploy them as statically linked libraries, as dynamically linked libraries -like cppmicroservices does it-, as different processes..).

some related terms which spit out good google results are flow based programming and dataflow programming (also going for composition of components)

Just created a pad: https://piratenpad.de/p/components

**Comment by IAmNotHanni** [comment]
Just for the record: the 3D Visual Scripting System runs every code execution (started by e.g. a timer) in an own thread, but I haven't thought of thread safety and critical sections/mutex yet.

**Comment by IAmNotHanni** [comment]
I've made the following illustration.

![Modules which will be based on the entity system](https://raw.githubusercontent.com/inexor-game/visualisations/8f77d5cc8ea6769205c249ed3a1173352415a095/multithreading/multithreading_strategy.png)

**Comment by a-teammate** [comment]
I think we can get a pretty cool design with [FastFlow](http://calvados.di.unipi.it/)

It is beginner-friendly (it avoids any deadlocks).
First task:
- [ ] find out which fastflow features we can apply for a specific scenario
   - can we use it for only specific parts of the engine?
